### PR TITLE
fix(appmesh): allow a Virtual Node have as a backend a Virtual Service whose provider is that Node

### LINKED ISSUE #17322: (AppMesh): How to create two virtual service/node with bidirectional backend without circular dependencies
### What is the problem?

I am currently setting up a CDK stack using ECS and AppMesh to create a micro service environment. In this micro service environment, every service should be able to talk to every other service. What I am trying to do is create a VirtualNode & Service per micro service and then call `addBackend` for each of these services and add all other services. However, this creates a cyclic dependencies for all services.

I haven't found a way yet to create this bidirectional backend setup without creating cyclic dependencies inside the CDK.

### Reproduction Steps

```ts
        // Inside 1 stack
        const node1 = new appmesh.VirtualNode(stack1, 'node1', {
            virtualNodeName: 'Node1',
            mesh: props.mesh,
            listeners: [appmesh.VirtualNodeListener.grpc({
                port: 8080,
                healthCheck: appmesh.HealthCheck.grpc({
                    healthyThreshold: 2,
                    interval: cdk.Duration.seconds(5),
                    timeout: cdk.Duration.seconds(2),
                    unhealthyThreshold: 3,
                }),
            })],
            accessLog: appmesh.AccessLog.fromFilePath('/dev/stdout'),
        });

        const service1 = new appmesh.VirtualService(stack2, 'Service1', {
            virtualServiceName:  'Service1',
            virtualServiceProvider: appmesh.VirtualServiceProvider.virtualNode(node1),
        });

       // Inside stack 2
        const node2 = new appmesh.VirtualNode(this, 'node2', {
            virtualNodeName: 'Node2',
            mesh: props.mesh,
            listeners: [appmesh.VirtualNodeListener.grpc({
                port: 8080,
                healthCheck: appmesh.HealthCheck.grpc({
                    healthyThreshold: 2,
                    interval: cdk.Duration.seconds(5),
                    timeout: cdk.Duration.seconds(2),
                    unhealthyThreshold: 3,
                }),
            })],
            accessLog: appmesh.AccessLog.fromFilePath('/dev/stdout'),
        });

        const service2 = new appmesh.VirtualService(this, 'Service2', {
            virtualServiceName: 'Service2',
            virtualServiceProvider: appmesh.VirtualServiceProvider.virtualNode(node2),
        });

      // Added afterwards via Aspect
      node1.addBackend(service2);
      node2.addBackend(service1)
```

### What did you expect to happen?

I expected there to be a way to create a bidirectional connection between backend services.

### What actually happened?

It always creates a circular dependency and I can't see any escape hatch

### CDK CLI Version

1.130.0

### Framework Version

_No response_

### Node.js Version

9.0.0

### OS

MacOS BigSur

### Language

Typescript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by wplucinsky**
Addresses a circular dependency issue between Virtual Nodes and Virtual Services that works for Virtual Services created with a defined `virtualServiceName` and a randomly generated name. 

One such example of this problem was a case where a Virtual Node had a backend that is a Virtual Service whose provider was given as the same Virtual Node. This led to the Virtual Node being dependent on the creation of the Virtual Service, and the Virtual Service being dependent on the creation of the Virtual Node.

Fixes #17322

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-appmesh/test/virtual-node.test.ts [hunk 1]
```diff
@@ -478,8 +470,51 @@ describe('virtual node', () => {
             ],
           },
         });
+      });
+
+      test('should add a backend virtual service to the resource without using a CDK Ref', () => {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/virtual-node.test.ts:475]: This name could use some work I think:

```suggestion
      test('you can add a Virtual Service as a backend to a Virtual Node which is the provider for that Service', () => {
```

File: packages/@aws-cdk/aws-appmesh/test/virtual-node.test.ts [hunk 2]
```diff
@@ -478,8 +470,51 @@ describe('virtual node', () => {
...
+          serviceDiscovery: appmesh.ServiceDiscovery.dns('test'),
+        });
 
+        const myVirtualService1 = new appmesh.VirtualService(stack, 'service-1', {
+          virtualServiceProvider: appmesh.VirtualServiceProvider.virtualNode(node),
+        });
 
+        const myVirtualService2 = new appmesh.VirtualService(stack, 'service-2', {
+          virtualServiceProvider: appmesh.VirtualServiceProvider.virtualNode(node),
+          virtualServiceName: 'service2.domain.local',
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/virtual-node.test.ts:495]: Can we not pass `virtualServiceName` here? I don't think we have a test yet that covers that the automated Virtual Service naming works with `Backend.virtualService()`.

**Review by skinny85 — CHANGES_REQUESTED** [review]
Looks good @wplucinsky! A few minor comments.

Additionally, the build is failing - looks like you need to update an integ test in the `@aws-cdk-containers/ecs-service-extensions` module:

```
@aws-cdk-containers/ecs-service-extensions: Verifying integ.all-service-addons.js against integ.all-service-addons.expected.json ... CHANGED.
```

File: packages/@aws-cdk/aws-appmesh/test/virtual-node.test.ts [hunk 3]
```diff
@@ -478,8 +470,51 @@ describe('virtual node', () => {
...
+        });
+
+        node.addBackend(appmesh.Backend.virtualService(myVirtualService1));
+        node.addBackend(appmesh.Backend.virtualService(myVirtualService2));
```
[the remark this answers is not in the document]
Comment (wplucinsky) [reply packages/@aws-cdk/aws-appmesh/test/virtual-node.test.ts:499]: I'll break it into two tests in the new commit, but one virtual service does not supply a `virtualServiceName`, and the other is when a `virtualServiceName` is supplied. That tests the two cases in the ternary `(this.virtualService as any).physicalName` and  `this.virtualService.virtualServiceName`.

File: packages/@aws-cdk/aws-appmesh/test/virtual-node.test.ts (hunk 2 above)
Comment (wplucinsky) [reply packages/@aws-cdk/aws-appmesh/test/virtual-node.test.ts:495]: That's covered by `myVirtualService1`, but I'll break them into two different tests.

**Comment by Seiya6329** [comment]
Thanks @wplucinsky and @skinny85! These are looking great!

**Comment by Seiya6329** [comment]
@wplucinsky - This is the friendly reminder that the build is failing for the `@aws-cdk-containers/ecs-service-extensions` failure.

**Review by skinny85 — APPROVED** [review]
Looks perfect, thanks @wplucinsky!

**Pull request merged**